Debug integration review (#1069)
riscv/riscv-cheri: src/cheri/debug-integration.adoc

Author: ```suggestion
* `hartinfo.datasize` is at least 1 and `hartinfo.dataaccess` is 0.
```

Author: ```suggestion
* `abstractcs.progbufsize` is at least 4 if `dmstatus.impebreak` is 1, or at least 5 if `dmstatus.impebreak` is 0.
```

Reviewer: not sure why that one got lost

Author: I would suggest to now drop the sentence: "These requirements may be relaxed if ...". That's because the preceding text now clearly says that the implementation should contain at least one of: abstract access for capabilities or a sufficiently configured program buffer.

```suggestion
These requirements allow a debugger to read and write capabilities in integer
registers without disturbing other registers.
```

Author: Depending on the outcome of the discussion under https://github.com/riscv/riscv-cheri/issues/1039, we may drop some of the sequences because they may become redundant and get replaced by the abstract commands.

Author: I think it would be useful to add program buffer sequences for reading / writing capabilities in memory to this non-normative note. 

However, given the amount of changes in this PR is already large, I would leave that for a subsequent PR.

Reviewer: agreed - I won't add them but I would be happy if you do!

=== Integrating {cheri_base_ext_name} with Sdext

{cheri_base_ext_name} harts must provide a means for accessing capabilities in registers and in memory via the external debugger.

This is achieved by supporting at least one of:

. Implementing a program buffer configured as specified below.
. Supporting full capability access for both Access Register and Access Memory abstract commands (`aarsize=7` and `aamsize=7`).

==== Program Buffer

If a Program Buffer is implemented, this section shows how it can be configured and used to access capabilities.

* `hartinfo` must be implemented.
* `hartinfo.nscratch` is at least 1 and the <<dscratch0_y>> register is implemented.
* `hartinfo.datasize` is at least 1 and `hartinfo.dataaccess` is 0.
* `abstractcs.progbufsize` is at least 4 if `dmstatus.impebreak` is 1, or at least 5 if `dmstatus.impebreak` is 0.

[NOTE]
====
These requirements allow a debugger to read and write capabilities in integer
registers without disturbing other registers.  These requirements may be
relaxed if some other means of accessing capabilities in integer registers,
These requirements allow a debugger to read and write capabilities in integer

> registers without disturbing other registers.

Author: This looks like a typo in an applied suggestion.

The resulting text of this paragraph should be just:

```
These requirements allow a debugger to read and write capabilities in integer
registers without disturbing other registers.
```

Reviewer: fixed

> {cheri_base_ext_name} harts should be able to access capabilities in registers and in memory via the external debugger.

Author: If this makes sense to you:

- `should -> must` - making it very clear that this is a requirement
- It is the external debugger who accesses the capabilities, and the harts only provide a way how to do it.

```suggestion
{cheri_base_ext_name} harts must provide a means how for accessing capabilities in registers and in memory via the external debugger.
```

Reviewer: ```suggestion
{cheri_base_ext_name} harts must provide a means for accessing capabilities in registers and in memory via the external debugger.
```

> This can be achieved by at least one of:

Author: How strong do we want this requirement to be?

- `can` - > `should`?
- `can` -> `shall`?
- ... ?

Reviewer: ```suggestion
This is achieved by supporting at least one of:
```

. Implementing a program buffer configured as specified below.

> . Supporting `aarsize=7` for abstract commands which access registers and memory, or both.

Author: - Is the second bullet point meant to say that aarsize=7/aamsize=7 must be supported for **both** Access Register and Access Memory abstract commands? If so, I would rephrase it to make it clearer.
- Does the final "or both" mean that the implementers are free to implement both the bullet points at the same time? If so, I would drop it because it is IMO redundant (we already said "at least one of") and not fully clear what it refers to.

```suggestion
This can be achieved by at least one of:

. Implementing a program buffer configured as specified below.
. Supporting full capability access for both Access Register and Access Memory abstract commands (`aarsize=7` and `aamsize=7`).
```

  setting `transfer` to 1 and `write` to 0).
====

==== Access Register Command

The _Access Register_ abstract command has optional support for accessing YLEN registers.

> If `aarsize` matches YLEN then all YLEN-bits should be accessible via the command.

Author: From the discussion under https://github.com/riscv/riscv-cheri/issues/1039 it so far looks to me that we should make YLEN access mandatory (not optional).

One possibility how to formulate that:

```suggestion
When an YLEN-wide register is accessible via _Access Register_ abstract command, then:

- Reads of YLEN bits or any smaller amounts of bits must be supported.
- Writes of YLEN bits must be supported.
- Writes of smaller amounts of bits than YLEN may optionally be supported but it is UNSPECIFIED what happens to the higher bits.

NOTE: The debugger can change the low XLEN bits of a YLEN wide register by performing a read-modify-write sequence.
```

Alternate possibility, which would optionally allow modify the low XLEN bits without altering the metadata in a single command:

```suggestion
When an YLEN-wide register is accessible via _Access Register_ abstract command, then:

- Reads of YLEN bits and any smaller amounts of bits must be supported.
- Writes of YLEN bits must be supported.
- Writes of XLEN bits may optionally be supported. They are supported, the capability metadata (the upper YLEN/2 bits of the register) must be left unchanged.
- Writes of less than XLEN bits may optionally be supported but it is UNSPECIFIED what happens to the higher bits.
```

This requires enough `data<n>` registers to be supported to cover the entire data width (2 for {cheri_base32_ext_name} or 4 for {cheri_base64_ext_name}).

Implementations can choose to also support accessing {ctag}s as part of the access register command.

`aarsize=7` is defined to represent full capability access.

The `data<n>` register mapping is as shown in <<debug_data_aarsize_7>>.

> .Data register mapping when `aarsize=7`

Author: To me as a reader, it was strange that this table is introduced in Access Register section but contains provisions for the Access Memory command (the arg1 column that holds memory addresses).

Could the table be moved to a generic section about abstract commands? 

Something like this:

```
==== Abstract Commands

<Common requirements go here, including the table with data registers.>

===== Access Register Command

<Specifics of access register command.>

===== Access Memory Command

<Specifics of access memory command.>
```

I think this would also make it easier to later re-insert the requirements back to the original text of the debug spec - if that is the eventual goal.

[#debug_data_aarsize_7]
[width="100%",options=header,cols="1,1,1,1,1"]
|=====

> |ISA                    | arg0/return value| arg1^1^     | arg2        | {ctag}

Author: The `arg2` column is strange - both the original debug spec and this extension contain it, but neither of the specs make any use of it.  🤔 

Should we still keep it - for consistency with the original debug spec? If yes, a note can be added, saying that the "arg2" is there for future extensions.

Reviewer: it seems to be just a aa[rm]size register

Author: <del>

I also think so, but still, the `arg2` argument is not utilized anywhere else in the RISC-V debug spec, so readers may wonder why it is even part of the table.

To avoid any doubts, we may add a note along these lines:

`NOTE: The argument "arg2" is currently not used by any abstract command and is left for future extensions.`

</del>

EDIT: `arg2` is mentioned in the non-normative "B.2.11. Quick Access" chapter as a possible temporary storage where to lay off data during program buffer sequences.

Reviewer: I'll add a note

Author: So given that, I think the table can stay as it is.

|{cheri_base32_ext_name}| data0,data1      | data2       | data3,data4 | data5
|{cheri_base64_ext_name}| data0-data3      | data4,data5 | data6-data9 | data10
|=====

^1^ arg1 represents an address so never needs to exceed DXLEN.
 Each `data<n>` register is 32-bit so a maximum of 2 is required.

Bit 0 of `data5` or `data10` contain the {ctag} which is either written to, or read from, the target register.
All other bits in that `data` register are UNSPECIFIED.

> When a capability is written it must be a <<cap_subset>> of the capability in <<drootc>>.

Author: Does this mean that the present value of `drootc` is relevant - as selected by `drootsel` at the time the abstract command is issued?

Reviewer: yes! for us that's not a problem as there's only one. CHERIoT would have to choose the correct root.

Author: Ok, understood.

Reviewer 2: This doesn't make sense with otype-based sealing. You require *multiple* roots in order to derive that capability: one to authorise the initial unsealed capability, and another to authorise the sealing.

Reviewer: are you saying that a written cap needs to be a subset of at least one of the possible root caps selectable by drootcsel? But I don't understand why a single written cap wouldn't be a subset of one of them.

Reviewer 2: If I want to write a capability that is `0x42 [rwRW,0x40-0x50]` but _sealed with otype 7_, say, and sealing capabilities have a separate root, what singular capability would exist in drootc to authorise it? I argue none, because to construct it with code you first need to YBLD the unsealed capability using the normal root of rwRW capabilities, and then YSEAL the result with the root of sealing capabilities.

Reviewer 2: Noting that (a) CHERIoT's multi-root scheme has rx, rw, and sealing as three distinct roots (b) we envision any otype-based sealing extension for "normal" Y as still having two distinct roots, rwx and sealing.

Reviewer: I guess I don't understand multiple roots well enough - in that case there's no root cap which can be used as an argument to YSS against the sealed cap which would pass?
so maybe it needs to be written unsealed and then sealed in registers before being written to memory.

Reviewer 2: > so maybe it needs to be written unsealed and then sealed in registers before being written to memory.

That sounds a lot like "don't use abstract commands, use the program buffer"? Which we're trying to get away from?

Reviewer 2: > in that case there's no root cap which can be used as an argument to YSS against the sealed cap which would pass?

Well, YSS would return 1 when given the appropriate unsealed root, because subset tests ignore any sealing. But a YBLD of that bit pattern (if untagged) wouldn't give you that sealed capability.

Reviewer: so what's your suggestion for fixing it so it works with multiple roots?

Reviewer 2: "Is derivable from the set of available debug root capabilities"?

Reviewer 2: (Probably with a non-normative note explaining this is "just" a subset check against each root for the "normal" Y base but extensions, or other bases like CHERIoT, that add more roots may require multiple check)

If not then the destination {ctag} is set to zero.

> NOTE: For the capability subset: `cap2` is the capability being written and `cap` 1 is the value of <<drootc>>.

Reviewer: ```suggestion
NOTE: For the capability subset: `cap2` is the capability being written and `cap1` is the value of <<drootc>>.
```

NOTE: It is valid for root capabilities not to cover the entire possible address space.
 In this case the debugger cannot claim access to more memory than is permitted by <<drootc>>.

> NOTE: Any register write with `aarsize{ne}7` causes the {ctag} (if present) to be set to zero.

Author: I would recommend to drop the "note" since this is a normative statement - requirement.

```suggestion
Any register write with `aarsize{ne}7` causes the {ctag} (if present) to be set to zero.
```

Reviewer: it's a note as this is following the standard rule for memory writing, as we don't repeat normative text.

Author: Ok, understood. In that case, let's keep the "NOTE" prefix.

==== Access Memory Command

> The _Access Memory_ abstract command supports accessing {ctag}s when `aarsize=7`, using the `data<n>` register format specified in <<debug_data_aarsize_7>>.

Author: 1) Could we use the formulation "the implementation can choose to support ..." also here to make it clear that the Access Memory command for a full capability is conditionally optional?

2) Typo:  The size field in Access Memory abstract commands is called `aamsize`.

```suggestion
Implementations can choose to support accessing full capabilities, including the {ctag}s, via the the _Access Memory_ abstract command. Full capability access is denoted by `aamsize=7` and uses the `data<n>` register assignments specified in <<debug_data_aarsize_7>>.
```

> 

Author: Should we state also here that the capability being written must be a subset of `drootc`, otherwise the tag is written as 0?

Reviewer: that's already above on line 122

Author: The line 122 is for Access Register abstract command, so I think the same requirement should be repeated for the Access Memory command, too. (Alternatively, the requirement can be moved to some form of common section for all abstract commands, if such a section is created.)

> NOTE: If `aarsize{ne}7` then {ctag}s are always read and written as zero.

Author: Similarly here - I would drop the "note" prefix.

```suggestion
If `aarsize{ne}7` then {ctag}s are always read and written as zero.
```

Reviewer: again, this follows the standard memory access pattern of tags not being accessed if the width is not YLEN

Author: Ok, understood.

|=====
|ISA                    | arg0/return value| arg1^1^     | arg2        | {ctag}
|{cheri_base32_ext_name}| data0,data1      | data2       | data3,data4 | data5
|{cheri_base64_ext_name}| data0-data3      | data4,data5 | data6-data9 | data10
|=====

^1^ arg1 represents an address so never needs to exceed DXLEN.
 Each `data<n>` register is 32-bit so a maximum of 2 are required.

^2^ arg2 is not used for abstract commands, and is included for consistency with the equivalent table in the _RISC-V Debug Specification_

Bit 0 of `data5` or `data10` contain the {ctag} which is either written to, or read from, the target register.
All other bits in that `data` register are UNSPECIFIED.

==== Access Register Command

When a YLEN-wide register is accessible via an _Access Register_ abstract command, then:

- Reads of YLEN bits or any smaller amounts of bits must be supported.
- Writes of YLEN bits must be supported.
- Writes of smaller amounts of bits than YLEN may optionally be supported but it is UNSPECIFIED what happens to the higher bits.

NOTE: The debugger can change the low XLEN bits of a YLEN wide register by performing a read-modify-write sequence.

> This requires enough `data<n>` registers to be supported to cover the entire data width (2 for {cheri_base32_ext_name} or 4 for {cheri_base64_ext_name}).

Author: Nitpicks: 

1) Blank line to visually separate the Note and the following paragraph.

2) Since the flow of the text is broken by the note, the "This requires ..." reference sounds somewhat odd to me. I tried to rephrase it below. 

3) I would say "at least 2 ... or at least 4 ..." instead of just "2 ... or 4 ...".

```suggestion
NOTE: The debugger can change the low XLEN bits of a YLEN wide register by performing a read-modify-write sequence.

The requirement to support YLEN-wide access mandates that enough `data<n>` registers are implemented to cover the entire YLEN bit-width (at least 2 for {cheri_base32_ext_name} or at least 4 for {cheri_base64_ext_name}).
```

Implementations can choose to also support accessing {ctag}s as part of the access register command by supporting `aarsize=7`.

When a capability is written it must be a <<cap_subset>> of the capability in <<drootc>>.
If not then the destination {ctag} is set to zero.

If `aarsize{ne}7` then {ctag}s are always read and written as zero.

NOTE: For the capability subset: `cap2` is the capability being written and `cap1` is the value of <<drootc>>.

NOTE: It is valid for root capabilities not to cover the entire possible address space.

>  In this case the debugger cannot claim access to more memory than is permitted by the current value of <<drootc>>.

Author: These requirements should be repeated for the Access Memory command, too. Or, even better, moved to the common section above.

Reviewer: done
